### ISSUE TITLE
Use one hierarchy of classes across the app to display lists of forms/instances

### ISSUE THREAD

**Issue by grzesiek2010** [getodk/collect#3756]
#### Problem description
Working on #799 I ran into tones of duplicated code. That stooped me from solving that issue.
This is how our code responsible for displaying lists of forms and instances look like now:
![Hierarchy-Page-1 (1)](https://user-images.githubusercontent.com/3276264/78236750-b9602580-74da-11ea-83ca-2ef47cca5441.png)
We have two very similar hierarchies one for fragments and one for activities.
Classes with the same color are almost identical but that's not the only problem since rest of them also contain a lot of duplicated code.

#### Expected behavior
The code should be refactored. We should use one hierarchy of classes. In order to do that we should use Fragments everywhere instead of activities.

**Comment by seadowg** [comment]
This game me some pain while working on Match Exactly. I think it'd be nice to take on as work towards 1.29. I think trying to move to Fragments everywhere for this would be great and as part of that I'm thinking pulling out shared view models and movign away from `AsyncTask` and `CursorLoader` would help us.

**Comment by grzesiek2010** [comment]
> This game me some pain while working on Match Exactly. I think it'd be nice to take on as work towards 1.29.

I agree this is not that important and we have had to many big prs recently so let's wait.

**Comment by seadowg** [comment]
> I agree this is not that important

Ah sorry if it came over that way! I really meant that "this seems even more important" 😩. Definitely not a priority for 1.28 as we can release with or without it but will be nice to start cleaning some of it up next release.

**Comment by grzesiek2010** [comment]
Ahh ok but now it is added to v1.28 so that's why I said we can wait (to avoid conflicts). Taking it just after releasing v1.28 seems fine.

**Comment by grzesiek2010** [comment]
Ok looks like now I should have time to tackle this issue.
Generally it's obvious that having two similar hierarchies (one for activities and one for fragments) is wrong and that's the one thing that should be fixed but when it comes to the hierarchy of classes itself (inheritance 3 or 4 levels of classes that we have now) I'm not that sure... I'm a different programmer than a yer ago when I reported this issue and I'm more into composition than inheritance now so I need to take another look at it to decide what's better (probably both but less inheritance) :smile: 

I think this shouldn't be only about removing that one hierarchy of classes. If we need to touch every list class either way I think it's a perfect time to add viewmodels to move the code out of activities/fragments, improve loading lists by removing the deprecated loaders https://github.com/getodk/collect/issues/3419, use recyclerviews instead of listviews etc.

If you have anything to add please feel free to do that.

**Comment by lognaturel** [comment]
Sounds great! A quick thing to add is that it makes sense to have the upcoming [projects support](https://forum.getodk.org/t/server-url-shown-on-home-screen-multiple-servers-switch-with-colour-change/23340) in mind. We'll presumably end up with an abstraction layer to get the correct form or instance list for the current project and it'd be great to take any opportunities there might be to pave the way for that.

**Comment by seadowg** [comment]
> I think this shouldn't be only about removing that one hierarchy of classes. If we need to touch every list class either way I think it's a perfect time to add viewmodels to move the code out of activities/fragments, improve loading lists by removing the deprecated loaders #3419, use recyclerviews instead of listviews etc.

Yeah, definitely liking what you're saying here. Pulling out the common components (`BlankFormListViewModel` feels like an example of this) should be a nice path to continue on here. Personally I think getting rid of loads would give us the most bang for our buck as that locks us into the `CursorLoader` stuff from the view code to the database code. I'm definitely happy to talk any of this through on Slack/calls when you're looking at it.

**Comment by grzesiek2010** [comment]
After investigating the issue I would like to share my thoughts.
As I said above when we built those hierarchies I was less experienced and more into inheritance, now I think it was wrong so merging those two hierarchies into one also wouldn't make our code better (more readable and testable).

First of all four classes from those hierarchies (`FormListActivity`/`InstanceListActivity`/`FormListFragment`/`InstanceListFragment`) exist solely to keep one method `getSortingOrder()` what doesn't make much sense and having a separate class, something like `ListSortingProvider`/`ListSorter` would be better. 

Second of all `AppListActivity` and `AppListFragment` which are parents for both hierarchies are parents for two different kinds of lists. Some lists are just clickable some of them support selecting forms in order to upload/delete/download them so in fact there is not much common behavior.

My first thought after investigating this issue again was that ok **let's first reduce those two hierarchies into one and we will see** but now I think it doesn't make sense because most likely we need separate classes and they won't need a common parent so I would do that the other way around: **let's improve our lists one by one in isolation introducing viewmodels/recyclerviews etc and then we will see if we need any parent.**

would love to know your opinion @seadowg @lognaturel

**Comment by seadowg** [comment]
> My first thought after investigating this issue again was that ok let's first reduce those two hierarchies into one and we will see but now I think it doesn't make sense because most likely we need separate classes and they won't need a common parent so I would do that the other way around: let's improve our lists one by one in isolation introducing viewmodels/recyclerviews etc and then we will see if we need any parent.

Yup this makes sense to me! I think it'll definitely be easier to understand after that.

**Comment by grzesiek2010** [comment]
In https://github.com/getodk/collect/pull/5077 I've refactored the list of blank forms. I've also looked into the whole hierarchy of lists that we have and now I have a bit different opinion. As I said in https://github.com/getodk/collect/pull/5077 it doesn't make sens to build one reusable component to display forms/instances. There is to little shared logic... instead we should try to factor out classes that can be shared indeed and use them in every single list (what I started doing in https://github.com/getodk/collect/pull/5077). That means we should close this issue.
